### PR TITLE
NEW Add template accessors and extension helper methods for getting locale data

### PR DESCRIPTION
**Pull request by robbieaverill**

### CODE REVIEW

**Comment by dhensby** [comment]
looks like this needs a rebase

**Comment by robbieaverill** [comment]
Thanks, updated

File: src/Extension/FluentExtension.php [hunk 1]
```diff
@@ -458,4 +458,40 @@ protected function getDataQueryLocale(DataQuery $dataQuery = null)
...
+            return $localeObj ? $localeObj->getBaseURL() : '';
+        }
+
+        // Mock locale and recalculate link
+        return FluentState::singleton()->withState(function ($newState) use ($locale) {
+            $newState->setLocale($locale);
+
+            $link = DataObject::get($this->owner->ClassName)->byID($this->owner->ID)->Link();
+
+            // Prefix with domain if available
```
Comment (tractorcow) [inline src/Extension/FluentExtension.php]: I'll have a look at this logic; We may need to check if we are in domain mode, first.

Alternatively this should probable be baked into the updateLink() extension method.

File: src/Middleware/DetectLocaleMiddleware.php [hunk 2]
```diff
@@ -186,8 +186,12 @@ protected function getPersistKey()
     protected function getParamLocale(HTTPRequest $request)
     {
         $queryParam = FluentDirectorExtension::config()->get('query_param');
-        $locale = (string)$request->param($queryParam) ?: $request->requestVar($queryParam);
-        return $locale;
+        if (FluentDirectorExtension::config()->get('detect_locale')
```
Comment (tractorcow) [inline src/Middleware/DetectLocaleMiddleware.php]: detect_locale is meant to be a flag for "make the guess guess if we don't have an explicit locale". In this case a query param is explicit so we don't need to treat it as though we would an ambiguous check.

File: src/Model/Locale.php [hunk 3]
```diff
@@ -296,4 +301,47 @@ public function getChain()
...
+     * Determine the base URL within the current locale
+     *
+     * @return string
+     */
+    public function getBaseURL()
+    {
+        $base = Director::baseURL();
+
+        // Build domain-specific base URL
+        if ($this->Domain() && $this->Domain()->exists()) {
```
Comment (tractorcow) [inline src/Model/Locale.php]: We should only do this in domain mode. Even if a domain is configured (but not in domain mode) we should default to normal locale-prefix baseurl.

File: src/Extension/FluentExtension.php [hunk 4]
```diff
@@ -511,4 +514,63 @@ protected function getManipulationRecordID($updates)
...
     }
+
+    /**
+     * Add the current locale's URL segment to the start of the URL
+     *
+     * @param  string &$base
+     * @param  string &$action
+     * @return null
+     */
+    public function updateRelativeLink(&$base, &$action)
```
Comment (tractorcow) [inline src/Extension/FluentExtension.php]: We should move this to FluentSiteTreeExtension, since it's sitetree specific.

Comment (tractorcow) [reply src/Extension/FluentExtension.php]: Also, we can use `updateLink` now, if you want to merge the domain with relativelink there. :)

**Comment by tractorcow** [comment]
I'm going to merge and make the changes in another PR.

**Pull request merged**